staking-contract/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+    }
+
+    /// Call to deposit money.
+    pub fn deposit(&mut self) {
+        self.ping();
+        let account_id = env::predecessor_account_id();
+        let user = if let Some(mut user) = self.users.get(&account_id) {
+            user.amount += env::attached_deposit();
+            user
+        } else {
+            User::new(&account_id, env::attached_deposit())
+        };
+        self.users.insert(&account_id, &user);
+    }
+
+    /// Stakes previously deposited money by given user on this account.
+    pub fn stake(&mut self, amount: String) {
+        let amount = u128::from_str(&amount).expect("Failed to parse amount");
+        self.ping();
+        let account_id = env::predecessor_account_id();
+        let mut user = self.users.get(&account_id).expect("User is missing");
+        user.stake(amount);
+        self.users.insert(&account_id, &user);
+////        println!("{:?} stake: {}, staked: {}, locked: {}", user, amount, self.staked_amount, env::account_locked_balance());
+        self.staked_amount += amount;
+        Promise::new(env::current_account_id()).stake(self.staked_amount, self.stake_public_key.clone());
+    }
+
+    /// Withdraws the non staked balance for given user.
+    pub fn withdraw(&mut self, amount: String) {
+        let amount = u128::from_str(&amount).expect("Failed to parse amount");
+        self.ping();
+        let account_id = env::predecessor_account_id();
+        let mut user = self.users.get(&account_id).expect("User is missing");
+        user.withdraw(amount);
+        self.users.insert(&account_id, &user);
+        Promise::new(account_id).transfer(amount);
+    }
+
+    /// Request withdrawal for epoch + 2 by sending unstaking transaction for
+    /// `current locked - (given user deposit + rewards)`
+    pub fn unstake(&mut self, amount: String) {
+        let amount = u128::from_str(&amount).expect("Failed to parse amount");
+        self.ping();
+        let account_id = env::predecessor_account_id();
+        let mut user = self.users.get(&account_id).expect("User is missing");
+        assert!(self.staked_amount >= amount);
+        user.unstake(amount);
+        self.users.insert(&account_id, &user);
+        self.staked_amount -= amount;
+//        println!("{:?} unstake {}, staked: {}, locked: {}", user, amount, self.staked_amount, env::account_locked_balance());
+        Promise::new(env::current_account_id()).stake(self.staked_amount, self.stake_public_key.clone());
+    }
+
+    /// Returns given user's liquid balance.
+    pub fn get_user_balance(&mut self, account_id: AccountId) -> String {
+        let user = self.users.get(&account_id).expect("User is missing");
+        user.amount.to_string()
+    }
+
+    /// Returns given user's staked balance.
+    pub fn get_user_stake(&mut self, account_id: AccountId) -> String {
+        let user = self.users.get(&account_id).expect("User is missing");
+        user.stake.to_string()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Add a test where the validator is slashed (but not all of the stake) and delegators unstake / widthraw what they have left.